src/radical/pilot/agent/scheduler/base.py

```diff
@@ -310,14 +314,33 @@ def create(cls, cfg, session):
     #
     # NOTE: any scheduler implementation which uses a different nodelist
     #       structure MUST overload this method.
+
+    # MING: This function is used to update the node list by taking in a list
+    #       of slots that have been assigned an update them to the new state.
+    #       The nodes are represented as a dictionary, where as the slots are
+    #       represented as a dictionary of a different structure. This confused
+    #       me at first when I tried to read it.
+
     def _change_slot_states(self, slots, new_state):
 
         for node_name, node_uid, cores, gpus in slots['nodes']:
 
             # Find the entry in the the slots list
+
+            # MING: Assuming 'uid' is the ID of the node, it seems a bit
+            #       wasteful to have to look at all of the nodes available
+            #       for use if at most one node can have that uid. Maybe
+            #       it would be worthwhile to simply keep a list of nodes
+            #       that we would read, and keep a dictionary that maps
+            #       the uid of the node to the location on the list?
+
             node = (n for n in self.nodes if n['uid'] == node_uid).next()
             assert(node)
 
+            # MING: This is a bit awkward to read since it node is the node
+            #       dictionary created during initialization, but cores 
+            #       refers to the core map associated with the slots. Can
+            #       be tricky to read at first. Same for GPUs
             for cslot in cores:
```

Comment:
Ack.  Any suggestion on how to improve?

Reply:
This depends on how we deal with how slots are changed, which is being discussed as well. We should come back to this once we've decided what should be done.